docs/manual/src/udl/decorators.md

```diff
@@ -0,0 +1,242 @@
+        } catch (e: Exception) {
+            errorReporter(e)
+        }
+    }
+
+    fun doConsequentialThing() {
+        demoRustObject.doConsequentialThing()
+        listener()
+    }
+}
+```
+
+This causes a proliferation of boiler plate code. Worse, everytime we add a new method to the `DemoRustObject`, handwritten foreign language code needs to be written to expose it safely.
+
+The more methods that want to do similar things, the more repetitive this gets.
+
+We could isolate the repeated code into a decorator class with `onBackgroundThread`, `withErrorReporter` and `thenNotifyListeners` methods.
+
+```kotlin
+class MyDemoDecorator(
+    val backgroundScope: CoroutineContext,
+    val errorReporter: (e: Exception) -> Unit,
+    val listener: () -> Void
+) {
+    fun onBackgroundThread(rustCall: () -> Unit) {
+        backgroundScope.launch {
+            rustCall()
+        }
+    }
+
+    fun <T> withErrorReporter(rustCall: () -> T) =
+        try {
+            rustCall()
+        } catch (e: Exception) {
+            errorReporter(e)
+        }
+
+    fun thenNotifyListeners(rustCall: () -> T) {
+        try {
+            rustCall()
+        } finally {
+            listener()
+        }
+    }
+}
+```
+
+Then, we can re-write the `DemoLib` as:
+
+```kotlin
+class DemoLib(
+    val demoRustObject: DemoRustObject,
+    val decorator: MyDemoDecorator,
+) {
+    fun doExpensiveThing() = decorator.onBackgroundThread {
+        demoRustObject.doExpensiveThing()
+    }
+
+    fun doCrashyThing() = decorator.withErrorReporter {
+        demoRustObject.doCrashyThing()
+    }
+
+    fun doConsequentialThing() = decorator.thenNotifyListeners() {
+        demoRustObject.doConsequentialThing()
+    }
+}
+```
+
+This is much better, but it's still looking a bit cut and pasty. Enter decorator objects.
+
+## Decorators to the rescue
+
+> Decorator objects take their name from [Python's decorator methods][py-decorators].
+>
+> In Pythonic terms, Uniffi's decorator objects are a collection of decorator functions.
+>
+> Swift and Kotlin don't provide functionality to capture arbitrary `*args` and call a function with those same `*args`, so
+> at this time, decorator objects aren't as powerful as Python decorators. Nevertheless, they can be still quite useful.
+
+[py-decorators]: https://www.python.org/dev/peps/pep-0318/#on-the-name-decorator
+
+### UDL
+
+In the UDL we:
+
+* Specify that `DemoLib` should be a decorated version of `DemoRustObject` using the `[Decorated=`] annotation
+* Specify how methods should by decorated using the `[CallsWith=]` annotation
+* Don't specify the class that holds the decorator functions.  This is done on a per-app basis in `uniffi.toml`.
+
+```webidl
+namespace demo {}
+
+[Decorated=DemoLib]
+interface DemoRustObject {
+    [CallsWith=onBackgroundThread]
+    void do_expensive_thing();
+
+    [Throws=RustyError, CallsWith=withErrorReporter]
+    void do_crashy_thing();
+
+    [CallsWith=thenNotifyListeners]
```

Comment:
It's kind of a shame that the required structure of the decorator class can't be declared in the UDL here for visibility (and for checking against typos etc) but I guess that's a function of UDL not being expressive enough for the type parameters and what-not.

Reply:
That's half of it.  The other half is you want flexibility for the consumers.  One consumer might want their `onBackgroundThread` decorator to return a `Promise<T>` and another might want a `Deferred<T>`.  One consumer might want the logger to handle exceptions and change the return type to `Option<T>` another might not handle exceptions and keep the return type the same.  Maybe there's a 3rd consumer that doesn't want to use decorators at all.

There's definitely a trade-off here, since you can't just have a single UDL file that declares everything.  When we were discussing wrapper/custom types it seemed like people were valuing the consumer flexibility, so I figured it would also be good here.

Reply:
> It's kind of a shame that the required structure of the decorator class can't be declared in the UDL here for visibility (and for checking against typos etc) but I guess that's a function of UDL not being expressive enough for the type parameters and what-not.

For completeness, #1051 is how that works out.